Flaky: TestServiceUpgrade_DecommissionCompleteAfterDataCleanup — decommissioned node still selected into truncation-point segment quorum, blocking auto-decommission

## Summary

`TestServiceUpgrade_DecommissionCompleteAfterDataCleanup` is flaky. It fails (120s timeout) when the **decommissioned node is still selected into the quorum of a segment created after decommission** — in particular the **truncation-point segment**, which the auditor never cleans. As a result the node keeps local segment data forever, `HasLocalSegmentData()` never returns `false`, and the decommission monitor never transitions the node to `decommissioned`.

This is **not** caused by any specific PR; it surfaced on a PR run but the relevant code paths (quorum selection, decommission monitor, auditor cleanup) are untouched by that PR. It is a pre-existing placement/gossip race.

## Failing CI

- Workflow run / job: https://github.com/zilliztech/woodpecker/actions/runs/28152345518/job/83372721184?pr=184
- Observed on PR #184, but reproducible independently of that PR's changes.

## Symptom

```
e2e_service_upgrade_test.go:439:
    Error:    Condition never satisfied
    Test:     TestServiceUpgrade_DecommissionCompleteAfterDataCleanup
    Messages: node should auto-transition to decommissioned after data cleanup
--- FAIL: TestServiceUpgrade_DecommissionCompleteAfterDataCleanup (130.65s)
```

Throughout the entire 120s `require.Eventually` window the target node (node0) stays at:

```
check: state=decommissioning, hasLocalData=true, remainingProcessors=18, safeToTerminate=false
```

All other `TestServiceUpgrade_Decommission*` cases in the same run PASS.

## Root cause (evidence from the run)

The test decommissions node0, sleeps 2s for gossip, writes more data to roll segments, then truncates at `segmentId=27`. It assumes post-decommission segments avoid node0 ("quorum selection now filters them out"), so the truncation-point segment lands on other nodes.

In this run that assumption did **not** hold:

1. node0 created writers for **18 segments**: `0,2,4,5,6,7,8,9,10,11,12,13,14,18,22,23,25,27`. Segments `18,22,23,25,27` were allocated at `07:03:52.2–52.8`, i.e. **after** the decommission + 2s gossip wait. (`remainingProcessors=18` matches this exactly.)
2. The truncation-point segment **27** was placed on a quorum that includes node0 — LAC was synced to `node=127.0.0.1:38045` (node0) for `segmentId=27`.
3. After `Truncate`, the auditor cleaned segments **0–26** (`successfulCleanups=27`) and node0's local `data.log` for those 17 segments was deleted (`deleteLocalFiles ... deletedCount=1`).
4. Segment **27 is the truncation point** (`truncatedSegmentId=27`, `minSegmentIdInUse=27`). The auditor **never cleans the truncation-point segment** by design, so node0's `.../node0/.../49/27/data.log` is never removed.
5. The decommission monitor only marks the node `decommissioned` when `HasLocalSegmentData() == false` (`server/service.go`). Since node0 still holds segment 27's `data.log`, `HasLocalSegmentData()` (`server/logstore.go`) keeps returning `true` → 120s timeout → FAIL.

So node0 created 18 segments, 17 were cleaned, segment 27 (truncation point) survived → permanent local data → decommission never completes.

## Why it is a race, not a deterministic bug

The test passes whenever quorum selection happens to keep the truncation-point segment (and all post-decommission segments) off the decommissioned node. It fails when the decommissioned node is still chosen. The 2s `time.Sleep` for gossip propagation plus the current node-selection path do not reliably exclude a `decommissioning` node.

## Proposed fixes

### Product side (the more fundamental fix)

Pick one (or both):

1. **Exclude `decommissioning`/`decommissioned` nodes from new-segment quorum selection reliably.** Today the test's premise (decommissioned node is filtered out of new placements) does not hold within a bounded time after `Decommission()`. The node-selection path (see `woodpecker/quorum/discovery.go` and the server-side pool/membership filtering it relies on) should honor the decommission tag so that no segment created after decommission lands on the node. This also avoids the wider problem that a decommissioned node can keep accreting new data.

2. **Make the decommission-completion check tolerant of legitimately-retained data.** `HasLocalSegmentData()` currently counts the truncation-point segment, which is intentionally never cleaned. If a decommissioned node holds only the protected truncation-point segment, it can never auto-complete. The completion check should not block on segments that are protected by the retention/truncation boundary (e.g. ignore segments with `segId >= minSegmentIdInUse`, or reconcile against metadata rather than raw on-disk `data.log` presence).

### Test side (make the test deterministic regardless of the race)

The test is intentionally trying to push the truncation point past the decommissioned node's data; it should assert that prerequisite instead of assuming it:

1. After decommission + gossip wait, **assert node0 is no longer selected** for new segments before continuing — e.g. wait until newly written segments' quorums no longer contain node0, then proceed. Fail fast with a clear message if filtering hasn't taken effect.
2. Alternatively, **verify the truncation-point segment's quorum does not include the decommissioned node** before `Truncate`; if it does, roll one more segment / retry placement so the truncation point lands off-node.
3. As a stopgap, increase/poll the gossip-propagation wait instead of a fixed 2s `time.Sleep`, and bound the test more tightly so a placement on the decommissioned node is detected and surfaced rather than silently timing out at 120s.

The test-side change makes CI green; the product-side change is what actually guarantees a decommissioned node drains and completes.

## References

- Test: `tests/integration/e2e_service_upgrade_test.go` (assertion at line 439)
- Decommission monitor / completion check: `server/service.go` (`decommissionMonitorLoop`, `GetDecommissionProgress`)
- Local data scan: `server/logstore.go` (`HasLocalSegmentData`)
- Auditor cleanup (skips truncation-point segment): `woodpecker/log/internal_log_writer.go`
- Quorum/node selection: `woodpecker/quorum/discovery.go`

--- a/tests/integration/e2e_service_upgrade_test.go
+++ b/tests/integration/e2e_service_upgrade_test.go
@@ -394,12 +394,12 @@ func TestServiceUpgrade_DecommissionCompleteAfterDataCleanup(t *testing.T) {
 	require.Equal(t, "decommissioning", targetNode.GetNodeStatus().State)
 	require.True(t, targetNode.GetDecommissionProgress().HasLocalData)
 
-	// Wait for gossip to propagate the decommission tag so new segments avoid this node
-	time.Sleep(2 * time.Second)
-
-	// Continue writing MORE data to force segment rolling. New segments will be placed
-	// on non-decommissioned nodes (quorum selection now filters them out). This causes
-	// the old active segment on the decommissioned node to be completed/fenced.
+	// Continue writing MORE data to force segment rolling. Once the "decommissioning" tag
+	// has propagated via gossip (see ServiceDiscovery.excludeDecommissioning), new segments
+	// are placed on non-decommissioned nodes; this completes/fences the old active segment
+	// on the decommissioned node. Gossip propagation is asynchronous, so some of these early
+	// segments may still land on the decommissioned node — that is fine, they get cleaned
+	// below once the truncation point advances past them.
 	const extraMsgs = 20
 	for i := 0; i < extraMsgs; i++ {
 		payload := make([]byte, msgSize)
@@ -414,36 +414,44 @@ func TestServiceUpgrade_DecommissionCompleteAfterDataCleanup(t *testing.T) {
 	}
 	t.Logf("Wrote %d additional messages after decommission to trigger segment rolling", extraMsgs)
 
-	// Write one final message to push data into a new segment beyond the decommissioned
-	// node's data. The auditor skips the truncation-point segment (segId >= truncatedSegmentId),
-	// so we need the truncation point to be in a segment AFTER all segments on the
-	// decommissioned node. This extra write ensures the last segment containing node3's
-	// data becomes eligible for cleanup.
-	finalPayload := make([]byte, msgSize)
-	finalResult := logWriter.Write(ctx, &log.WriteMessage{Payload: finalPayload})
-	require.NoError(t, finalResult.Err, "final post-decommission write failed")
+	// Drive the decommissioned node to drain by repeatedly advancing the truncation point.
+	//
+	// The auditor never cleans the truncation-point segment itself (it skips segId >= the
+	// truncation point). The original test wrote one final message and truncated once, then
+	// waited up to 120s for the node to drain. That raced with gossip convergence: if the
+	// truncation-point segment happened to land on the decommissioned node before its tag had
+	// propagated, that segment was never cleaned, so the node kept local data forever and
+	// never auto-transitioned to decommissioned (issue #203).
+	//
+	// Instead, keep rolling fresh segments and advancing the truncation point past them. Every
+	// segment below the truncation point is cleaned on all quorum nodes (including the
+	// decommissioned one); once gossip has converged, the new truncation-point segment lands
+	// off the decommissioned node, so its last remaining data is cleaned and it transitions to
+	// decommissioned. This is robust regardless of exactly when the decommission tag converges.
+	t.Log("Advancing truncation point until the decommissioned node drains...")
+	deadline := time.Now().Add(120 * time.Second)
+	drained := false
+	for time.Now().Before(deadline) {
+		// Roll at least one fresh segment (msgSize=1KB, 2KB segments) and truncate past it.
+		var truncateAt *log.LogMessageId
+		for i := 0; i < 3; i++ {
+			result := logWriter.Write(ctx, &log.WriteMessage{Payload: make([]byte, msgSize)})
+			require.NoError(t, result.Err, "drain-phase write failed")
+			truncateAt = result.LogMessageId
+		}
+		require.NoError(t, logHandle.Truncate(ctx, truncateAt))
 
-	// Truncate all data — mark everything as truncatable
-	truncateAt := finalResult.LogMessageId
-	t.Logf("Truncating at last entry: segmentId=%d, entryId=%d",
-		truncateAt.SegmentId, truncateAt.EntryId)
-	err = logHandle.Truncate(ctx, truncateAt)
-	require.NoError(t, err)
-
-	// Wait for: retention (2s) + auditor cycles (2s each) + cleanup execution
-	// The auditor will detect truncated segments, check retention TTL, and call CleanSegment
-	// on all quorum nodes including our decommissioning node.
-	// The decommission monitor will then detect no local data and mark decommissioned.
-	t.Log("Waiting for retention expiry + cleanup + decommission monitor...")
-
-	require.Eventually(t, func() bool {
 		state := targetNode.GetNodeStatus().State
 		progress := targetNode.GetDecommissionProgress()
-		t.Logf("  check: state=%s, hasLocalData=%v, remainingProcessors=%d, safeToTerminate=%v",
-			state, progress.HasLocalData, progress.RemainingProcessors, progress.SafeToTerminate)
-		return state == string(server.NodeStateDecommissioned)
-	}, 120*time.Second, 3*time.Second,
-		"node should auto-transition to decommissioned after data cleanup")
+		t.Logf("  check: state=%s, hasLocalData=%v, truncateAt=seg%d, remainingProcessors=%d, safeToTerminate=%v",
+			state, progress.HasLocalData, truncateAt.SegmentId, progress.RemainingProcessors, progress.SafeToTerminate)
+		if state == string(server.NodeStateDecommissioned) {
+			drained = true
+			break
+		}
+		time.Sleep(3 * time.Second)
+	}
+	require.True(t, drained, "node should auto-transition to decommissioned after data cleanup")
 
 	// Verify final state
 	progress := targetNode.GetDecommissionProgress()

